service/src/main/java/com/ge/predix/acs/policy/evaluation/cache/AbstractPolicyEvaluationCache.java

```diff
@@ -150,139 +176,136 @@ public void reset(final PolicyEvaluationRequestCacheKey key) {
                 .collect(Collectors.toList()));
     }
 
     private void multisetForResources(final String zoneId, final List<String> resourceIds) {
         String timestamp = timestampValue();
 
-        List<String> fromKeys = resourceIds.stream().map(resource -> resourceTranslationKey(zoneId, resource))
-                .collect(Collectors.toList());
-
-        List<Object> toKeys = multiGetResourceTranslations(fromKeys);
-
         Map<String, String> map = new HashMap<>();
-        for (int i = 0; i < toKeys.size(); i++) {
-            @SuppressWarnings("unchecked")
-            Set<String> toKeySet = (Set<String>) toKeys.get(i);
-            for (String toKey : toKeySet) {
-                map.put(toKey, timestamp);
-                String resourceId = resourceIds.get(i);
-                logSetResourceTimestampsDebugMessage(timestamp, toKey, resourceId);
-            }
+
+        for (String resourceId : resourceIds) {
+            String resourceKey = resourceKey(zoneId, resourceId);
+            map.put(resourceKey, timestamp);
+            logSetEntityTimestampsDebugMessage(timestamp, resourceKey, resourceId, EntityType.RESOURCE);
         }
         multiSet(map);
     }
 
-    private void logSetResourceTimestampsDebugMessage(final String timestamp, final String toKey,
-            final String resourceId) {
-        if (LOGGER.isDebugEnabled()) {
-            LOGGER.debug(String.format("Setting timestamp for resource '%s'; key: '%s', value: '%s' ", resourceId,
-                    toKey, timestamp));
-        }
+    @Override
+    public void resetForSubject(final String zoneId, final String subjectId) {
+        resetForEntity(zoneId, subjectId, EntityType.SUBJECT, AbstractPolicyEvaluationCache::subjectKey);
     }
 
     @Override
-    public void resetForSubject(final String zoneId, final String subjectId) {
-        String key = subjectKey(zoneId, subjectId);
-        String timestamp = timestampValue();
-        logSetSubjectTimestampDebugMessage(key, timestamp, subjectId);
-        set(key, timestamp);
+    public void setSubjectIfNotExists(final String zoneId, final String subjectId) {
+        setEntityIfNotExists(zoneId, subjectId, AbstractPolicyEvaluationCache::subjectKey);
     }
 
     @Override
     public void resetForSubjectsByIds(final String zoneId, final Set<String> subjectIds) {
-        multisetForSubjects(zoneId, subjectIds.stream().collect(Collectors.toList()));
+        multisetForSubjects(zoneId, new ArrayList<>(subjectIds));
     }
 
     @Override
     public void resetForSubjects(final String zoneId, final List<SubjectEntity> subjectEntities) {
         multisetForSubjects(zoneId,
-                subjectEntities.stream().map(subject -> subject.getSubjectIdentifier()).collect(Collectors.toList()));
+                subjectEntities.stream().map(SubjectEntity::getSubjectIdentifier).collect(Collectors.toList()));
     }
```

Comment:
Same comment as for resetForResources above.  Could we combine resetForSubjects and multisetForSubjects in one long beautiful fluent java 8 call? :-)